lalrpop/src/normalize/precedence/mod.rs

```diff
@@ -0,0 +1,334 @@
+/// ```
+///
+/// ## None
+///
+/// Non-associativity means that it is not legal to iterate the rule, turning our example to
+/// a parsing error. In this case, all recursive occurrences of the current rule are replaced with
+/// the rule corresponding to the previous level:
+///
+/// ```
+/// <left: Expression2> "?" <middle: Expression2> : <right: Expression3> => ..
+/// ```
+///
+/// ## Default
+///
+/// By default, if no associativity is provided, all recursive occurrences are replaced with the
+/// current level, which is different from non-associativity. This can be useful for unary
+/// operators that may be iterated, such as `-` or `!`.
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub enum Assoc {
+    Left,
+    Right,
+    NonAssoc,
+}
+
+/// Substitution plan.
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub enum Substitution<'a> {
+    /// Replace the first encountered occurrence by the first argument, and all the following by
+    /// the second. Used for associativity: typically, a left associativity on level `3` perform a
+    /// `OneThen(Rule3, Rule2)`.
+    OneThen(&'a SymbolKind, &'a SymbolKind),
+    /// Standard substitution mode. Replace every encountered occurrence with the same given
+    /// symbol.
+    Every(&'a SymbolKind),
+}
+
+/// Direction for substitution.
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub enum Direction {
+    Forward,
+    Backward
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseAssocError {
+    _priv: (),
+}
+
+impl fmt::Display for ParseAssocError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        "provided value was neither `left`, `right` nor `none`".fmt(f)
+    }
+}
+
+impl FromStr for Assoc {
+    type Err = ParseAssocError;
+
+    fn from_str(s: &str) -> Result<Assoc, ParseAssocError> {
+        match s {
+            "left" => Ok(Assoc::Left),
+            "right" => Ok(Assoc::Right),
+            "none" => Ok(Assoc::NonAssoc),
+            _ => Err(ParseAssocError { _priv: () }),
+        }
+    }
+}
+
+/// Perform precedence expansion. Rewrite rules where at least one alternative have a precedence
+/// annotation, and generate derived rules for each level of precedence.
+pub fn expand_precedence(input: Grammar) -> NormResult<Grammar> {
+    let input = resolve::resolve(input)?;
+
+    let items = input.items;
+
+    // Filter non terminals that have at least one precedence annotation
+    let (non_terms_prec, items): (Vec<_>, Vec<_>) = items
+        .into_iter()
+        .partition(|input| match input {
+            GrammarItem::Nonterminal(d) => has_prec_annot(d),
+            _ => false
+        });
+
+    let nonterms_prec: NormResult<Vec<Vec<GrammarItem>>> = non_terms_prec.into_iter().map(|item| match item {
+        GrammarItem::Nonterminal(d) => expand_nonterm(d),
+        _ => panic!()
+    }).collect();
+    let mut result: Vec<GrammarItem> = nonterms_prec?.into_iter().flatten().collect();
```

Comment:
This will change the order of the items which could be confusing. Is there a reason for this or could it be preserved?

Reply:
You're right, I think it's just because I started from the code for macro expansion, but macro are eliminated, so it doesn't matter there, as opposed as here. I'll make the transformation order preserving.

Reply:
Fixed in https://github.com/lalrpop/lalrpop/pull/555/commits/da2f17275ec96c2b229884bf767baf5eabf1f5b6